### PR TITLE
Add integration test for erroring when memory limits are hit

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it is waiting on:~
- [x] Builds on https://github.com/apache/arrow-datafusion/pull/4405

# Which issue does this PR close?

Resolves https://github.com/apache/arrow-datafusion/issues/4404

# Rationale for this change

I would like end to end coverage that when run with a memory limit, queries will error (rather than exceed the memory limit)

# What changes are included in this PR?
1. Add new `memory_limit` test
2. Improve error messages to make it clearer what is out of memory
2. Refactor `stagger_batch` into a function

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/execution/memory_manager/proxy.rs [hunk 1]
```diff
@@ -88,9 +88,10 @@ impl MemoryConsumer for MemoryConsumerProxy {
     }
 
     async fn spill(&self) -> Result<usize, DataFusionError> {
-        Err(DataFusionError::ResourcesExhausted(
-            "Cannot spill AggregationState".to_owned(),
-        ))
+        Err(DataFusionError::ResourcesExhausted(format!(
+            "Cannot spill {}",
```
Comment (alamb) [inline datafusion/core/src/execution/memory_manager/proxy.rs:92]: Improved error message here

File: datafusion/core/src/physical_plan/sorts/sort.rs [hunk 2]
```diff
@@ -272,6 +273,13 @@ impl MemoryConsumer for ExternalSorter {
...
 
     async fn spill(&self) -> Result<usize> {
+        let partition = self.partition_id();
+        let mut in_mem_batches = self.in_mem_batches.lock().await;
+        // we could always get a chance to free some memory as long as we are holding some
+        if in_mem_batches.len() == 0 {
+            return Ok(0);
+        }
+
         debug!(
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sorts/sort.rs:283]: I just moved the debug message down as it was confusing that spill didn't actually happen if `in_mem_batches` was empty

File: datafusion/core/tests/join_fuzz.rs [hunk 3]
```diff
@@ -200,24 +199,14 @@ fn make_staggered_batches(len: usize) -> Vec<RecordBatch> {
...
         ("b", Arc::new(input2) as ArrayRef),
         ("x", Arc::new(input3) as ArrayRef),
         ("y", Arc::new(input4) as ArrayRef),
     ])
     .unwrap();
 
-    let mut batches = vec![];
-
     // use a random number generator to pick a random sized output
-    let mut rng = StdRng::seed_from_u64(42);
```
Comment (alamb) [inline datafusion/core/tests/join_fuzz.rs:214]: This piece of code was copied in several places so I refactored it into `stagger_batch`

**Comment by alamb** [comment]
cc @crepererum  and @milenkovicm

File: datafusion/core/tests/memory_limit.rs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+async fn group_by_hash() {
+    run_limit_test(
+        // group by dict column
+        "select count(*) from t GROUP BY service, host, pod, container",
+        "Resources exhausted: Cannot spill GroupBy Hash Accumulators",
+    )
+    .await
+}
+
+/// 100K memory limit
```
Comment (crepererum) [inline datafusion/core/tests/memory_limit.rs:74]: 100k? How does relate to 50 and 0.95?

**Review by crepererum — APPROVED** [review]
Nice 💪

File: datafusion/core/tests/memory_limit.rs (hunk 4 above)
Comment (alamb) [reply datafusion/core/tests/memory_limit.rs:74]: Whoops -- the 100k is left over from a previous revision -- I will update the test

Comment (alamb) [inline datafusion/core/tests/memory_limit.rs:74]:
```suggestion
/// 50 byte memory limit
```

**Pull request merged**